lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -903,35 +905,117 @@ def test_filestream(
-            m.setattr(Controller, "__init__", TestCacheManager.MockController.__init__)
             m.setattr(tarfile, "open", fake_tarfile_open)
-            tb = Tarball(tar, Controller(Path("/mock/archive"), cache, None))
-            tar_dir = TestCacheManager.MockController.generate_test_result_tree(
-                tmp_path, "dir_name"
-            )
-            tb.cache_map(tar_dir)
 
-            expected_error_msg = (
-                f"The dataset tarball named '{tb.tarball_path}' is not found"
-            )
+            expected_error_msg = f"The dataset tarball named '{tar}' is not found"
             with pytest.raises(TarballNotFound) as exc:
-                tb.filestream("subdir1/f11.txt")
+                Tarball.extract(tar, Path("subdir1/f11.txt"))
+            assert str(exc.value) == expected_error_msg
+
+    def test_tarfile_extractfile_fails(self, monkeypatch, tmp_path):
+        """Test to check non-existent path in tarfile"""
+        tar = Path("/mock/result.tar.xz")
+        path = Path("subdir/f11.txt")
+
+        class MockTarFile:
+            def extractfile(self, path):
+                raise Exception("Mr Robot refuses trivial human command")
+
+        def fake_tarfile_open(self, path):
+            return MockTarFile()
+
+        with monkeypatch.context() as m:
+            m.setattr(tarfile, "open", fake_tarfile_open)
+            expected_error_msg = f"Unable to extract {path} from {tar.name}"
+            with pytest.raises(CacheExtractBadPath) as exc:
+                Tarball.extract(tar, path)
             assert str(exc.value) == expected_error_msg
 
+    def test_tarfile_extractfile_notfile(self, monkeypatch, tmp_path):
+        """Test to check target that's not a file"""
+        tar = Path("/mock/result.tar.xz")
+        path = Path("subdir/f11.txt")
+
+        class MockTarFile:
+            def extractfile(self, path):
+                return None
+
+        def fake_tarfile_open(self, path):
+            return MockTarFile()
+
+        with monkeypatch.context() as m:
+            m.setattr(tarfile, "open", fake_tarfile_open)
+            expected_error_msg = f"Unable to extract {path} from {tar.name}"
+            with pytest.raises(CacheExtractBadPath) as exc:
+                Tarball.extract(tar, path)
+            assert str(exc.value) == expected_error_msg
+
+    @pytest.mark.parametrize(
+        "tarball,stream", (("hasmetalog.tar.xz", True), ("nometalog.tar.xz", False))
+    )
+    def test_get_metadata(self, monkeypatch, tarball, stream):
+        """Verify access and processing of `metadata.log`"""
+
+        @staticmethod
+        def fake_extract(t: Path, f: Path):
+            if str(t) == tarball:
+                if str(f) == f"{Dataset.stem(t)}/metadata.log":
+                    if stream:
+                        return io.BytesIO(b"[test]\nfoo = bar\n")
+                    else:
+                        raise CacheExtractBadPath(t, f)
```

Comment:
I was tempted to ignore this as unnecessary noise, until I realize that we both failed to notice that we're returning a byte stream rather than an `Inventory` here. And as long as I'm hacking, I might as well take the bigger swing ...